README.md
General
=======

Two Factor Keeper is a basic PyQt5 GUI wrapper around the [oathtool commandline
program](http://www.nongnu.org/oath-toolkit/oathtool.1.html), and allows the
user quick access to any number of oath/2FA/2 Factor Authentication credentials,
of which the keys/secrets are protected by the [3rd party Python cryptography package Fernet
implementation](https://cryptography.io/en/latest/fernet/#using-passwords-with-fernet),
specific implementation details [here](https://cryptography.io/en/latest/fernet/#implementation).

This is my first PyQt5/Qt Designer project.

**Note that neither this program nor the cryptography library used have been
audited by a cryptography expert, so they are only assumed to be secure.** Use
at your own risk, unfortunately (security limitations with this library are
touched on [here](https://cryptography.io/en/latest/limitations/)).

**I currently only have TOTP sites in use with this, so HOTP is untested but
should work fine as oathtool is the thing doing the real work. Please report
your success with HOTP!**


Dependencies
============

Coded with/dependent on (may work with earlier versions):

Python 3.5  
PyQt5 5.7  
python3-cryptography 1.7.1  
python3-sh 1.11  
python3-yaml 3.12  
oathtool 2.6.1

All 3rd party dependencies are available packaged in Debian/Devuan.


Installation
============

As of yet there is no need to install this script, as long as the dependencies
are satisfied and the 'ui_mainwindow.py' script accompanies
'two-factor-keeper.py' - you can run the latter anywhere.


Usage
=====

Create Your First One Time Password Configuration Slot
------------------------------------------------------

On the first run of Two Factor Keeper, you'll be met with a mostly blank
interface:

![First run UI](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/first-run-ui.png?raw=true)

2FA configuration/credentials are collected into 'slots', and the left empty
list indicates there are no slots configured yet.

On the right, type a name that you will remember to identify the credentials
(e.g. site name), then enter the key/secret that the service/site has given you
(make sure there are no empty lines or spaces at the end). Usually this is a
base32 encoded string (just a long string of capital letters and numbers
potentially with one or more '=' at the end) - this option is on by default.

Encrypt the secret via the top-right lock button (currently in the unlocked
state), and enter your passphrase:

![Encrypt OTP secret](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/encrypt-otp-secret.png?raw=true)

Once a valid passphrase has been accepted, the key/secret widget locks, and the
encryption button has now changed to a locked decryption button:

![OTP secret encrypted](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/otp-secret-encrypted.png?raw=true)

Under 'Mode and basic options', choose TOTP (Time-based One Time Password)
rather than HOTP (Hash-based One Time Password) if the site you're configuring
for uses Google-style 2FA. If this is the case, all other options will work
with the defaults (SHA1 TOTP algorithm etc).

For sites that support it, you can configure more secure OTP generation with
more advanced hashes resulting in a number up to 8 digits long (the current
oathtool limit).

Further detail on the advanced options is available in the [oathtool manpage](http://www.nongnu.org/oath-toolkit/oathtool.1.html).


Generating A One Time Password
------------------------------

Once you are happy with the configuration, to create the first slot you can use
either the Apply or Add buttons at the bottom.

To generate the OTP, press the 'Generate OTP' button on the bottom left -
you'll be prompted for the passphrase:

![Generate OTP secret passphrase prompt](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/generate-otp-secret-passphrase-prompt.png?raw=true)

Get it right, and your OTP code is ready:

![OTP generated](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/otp-generated.png?raw=true)

This code is usually valid for 30 seconds after the site first asks for it, and
for your convenience it has also been copied to the clipboard.


Further Configuration
---------------------

*Slot changes are not saved automatically - make your change then press the Apply
button at the bottom to save.*

To change the key/secret, press the locked button and enter the passphrase - if
you no longer remember the passphrase, and don't mind losing the secret, press
the brush button to the right of the lock to erase:

![Erase OTP secret](https://github.com/OmegaPhil/two-factor-keeper/blob/master/docs/erase-otp-secret.png?raw=true)

If you make a change to a slot, haven't saved yet and want to undo it, use the
Reset button at the bottom to load the previously-saved configuration.

To remove a slot, select it in the list then use the Remove button on the bottom
right.


Adding More Slots
-----------------

Edit the current loaded slot and change the name, then instead of Applying any
changes, use the Add button.


Configuration Storage
=====================

Slot configuration is maintained in the '~/.config/two-factor-keeper/slots.conf'
YAML file.


Bugs And Feature Requests
=========================

Please create an issue on the [Github issue tracker](https://github.com/OmegaPhil/two-factor-keeper/issues).


Contact Details
===============

OmegaPhil: OmegaPhil@startmail.com


NEWS.md
v0.1 (6.06.17): First public release - TOTP in regular use, HOTP untested.


ui_mainwindow.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'mainwindow.ui'
#
# Created by: PyQt5 UI code generator 5.7
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore, QtGui, QtWidgets

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(638, 682)
        icon = QtGui.QIcon.fromTheme("changes-prevent")
        MainWindow.setWindowIcon(icon)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.MinimumExpanding, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.centralwidget.sizePolicy().hasHeightForWidth())
        self.centralwidget.setSizePolicy(sizePolicy)
        self.centralwidget.setObjectName("centralwidget")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.centralwidget)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.gridLayout = QtWidgets.QGridLayout()
        self.gridLayout.setObjectName("gridLayout")
        self.advanced_groupbox = QtWidgets.QGroupBox(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.advanced_groupbox.sizePolicy().hasHeightForWidth())
        self.advanced_groupbox.setSizePolicy(sizePolicy)
        self.advanced_groupbox.setObjectName("advanced_groupbox")
        self.gridLayout_3 = QtWidgets.QGridLayout(self.advanced_groupbox)
        self.gridLayout_3.setObjectName("gridLayout_3")
        self.totp_start_time_lineedit = QtWidgets.QLineEdit(self.advanced_groupbox)
        self.totp_start_time_lineedit.setEnabled(False)
        self.totp_start_time_lineedit.setObjectName("totp_start_time_lineedit")
        self.gridLayout_3.addWidget(self.totp_start_time_lineedit, 8, 0, 1, 2)
        self.label_5 = QtWidgets.QLabel(self.advanced_groupbox)
        self.label_5.setObjectName("label_5")
        self.gridLayout_3.addWidget(self.label_5, 5, 0, 1, 1)
        self.totp_time_step_spinbox = QtWidgets.QSpinBox(self.advanced_groupbox)
        self.totp_time_step_spinbox.setEnabled(False)
        self.totp_time_step_spinbox.setMinimum(1)
        self.totp_time_step_spinbox.setMaximum(9999)
        self.totp_time_step_spinbox.setProperty("value", 30)
        self.totp_time_step_spinbox.setObjectName("totp_time_step_spinbox")
        self.gridLayout_3.addWidget(self.totp_time_step_spinbox, 6, 1, 1, 1)
        self.label_7 = QtWidgets.QLabel(self.advanced_groupbox)
        self.label_7.setObjectName("label_7")
        self.gridLayout_3.addWidget(self.label_7, 7, 0, 1, 1)
        self.hotp_counter_spinbox = QtWidgets.QSpinBox(self.advanced_groupbox)
        self.hotp_counter_spinbox.setEnabled(True)
        self.hotp_counter_spinbox.setObjectName("hotp_counter_spinbox")
        self.gridLayout_3.addWidget(self.hotp_counter_spinbox, 5, 1, 1, 1)
        self.label_8 = QtWidgets.QLabel(self.advanced_groupbox)
        self.label_8.setObjectName("label_8")
        self.gridLayout_3.addWidget(self.label_8, 9, 0, 1, 1)
        self.label_6 = QtWidgets.QLabel(self.advanced_groupbox)
        self.label_6.setObjectName("label_6")
        self.gridLayout_3.addWidget(self.label_6, 6, 0, 1, 1)
        self.totp_now_lineedit = QtWidgets.QLineEdit(self.advanced_groupbox)
        self.totp_now_lineedit.setEnabled(False)
        self.totp_now_lineedit.setObjectName("totp_now_lineedit")
        self.gridLayout_3.addWidget(self.totp_now_lineedit, 10, 0, 1, 2)
        self.gridLayout.addWidget(self.advanced_groupbox, 2, 1, 1, 1)
        self.general_groupbox = QtWidgets.QGroupBox(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.general_groupbox.sizePolicy().hasHeightForWidth())
        self.general_groupbox.setSizePolicy(sizePolicy)
        self.general_groupbox.setObjectName("general_groupbox")
        self.gridLayout_4 = QtWidgets.QGridLayout(self.general_groupbox)
        self.gridLayout_4.setObjectName("gridLayout_4")
        self.name_lineedit = QtWidgets.QLineEdit(self.general_groupbox)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.name_lineedit.sizePolicy().hasHeightForWidth())
        self.name_lineedit.setSizePolicy(sizePolicy)
        self.name_lineedit.setObjectName("name_lineedit")
        self.gridLayout_4.addWidget(self.name_lineedit, 1, 0, 1, 1)
        self.label_2 = QtWidgets.QLabel(self.general_groupbox)
        self.label_2.setObjectName("label_2")
        self.gridLayout_4.addWidget(self.label_2, 0, 0, 1, 1)
        self.base32_encoded_checkbox = QtWidgets.QCheckBox(self.general_groupbox)
        self.base32_encoded_checkbox.setChecked(True)
        self.base32_encoded_checkbox.setObjectName("base32_encoded_checkbox")
        self.gridLayout_4.addWidget(self.base32_encoded_checkbox, 7, 0, 1, 1)
        self.widget_2 = QtWidgets.QWidget(self.general_groupbox)
        self.widget_2.setObjectName("widget_2")
        self.horizontalLayout_4 = QtWidgets.QHBoxLayout(self.widget_2)
        self.horizontalLayout_4.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout_4.setObjectName("horizontalLayout_4")
        self.label_3 = QtWidgets.QLabel(self.widget_2)
        self.label_3.setObjectName("label_3")
        self.horizontalLayout_4.addWidget(self.label_3)
        spacerItem = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.horizontalLayout_4.addItem(spacerItem)
        self.encrypt_pushbutton = QtWidgets.QPushButton(self.widget_2)
        self.encrypt_pushbutton.setText("")
        icon = QtGui.QIcon.fromTheme("changes-allow")
        self.encrypt_pushbutton.setIcon(icon)
        self.encrypt_pushbutton.setObjectName("encrypt_pushbutton")
        self.horizontalLayout_4.addWidget(self.encrypt_pushbutton)
        self.clear_key_pushbutton = QtWidgets.QPushButton(self.widget_2)
        self.clear_key_pushbutton.setText("")
        icon = QtGui.QIcon.fromTheme("edit-clear")
        self.clear_key_pushbutton.setIcon(icon)
        self.clear_key_pushbutton.setObjectName("clear_key_pushbutton")
        self.horizontalLayout_4.addWidget(self.clear_key_pushbutton)
        self.gridLayout_4.addWidget(self.widget_2, 5, 0, 1, 1)
        self.key_plaintextedit = QtWidgets.QPlainTextEdit(self.general_groupbox)
        self.key_plaintextedit.setObjectName("key_plaintextedit")
        self.gridLayout_4.addWidget(self.key_plaintextedit, 6, 0, 1, 1)
        self.gridLayout.addWidget(self.general_groupbox, 0, 1, 1, 1)
        self.otp_list = QtWidgets.QListWidget(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Preferred, QtWidgets.QSizePolicy.Expanding)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.otp_list.sizePolicy().hasHeightForWidth())
        self.otp_list.setSizePolicy(sizePolicy)
        self.otp_list.setMinimumSize(QtCore.QSize(0, 0))
        self.otp_list.setObjectName("otp_list")
        self.gridLayout.addWidget(self.otp_list, 0, 0, 3, 1)
        self.mode_groupbox = QtWidgets.QGroupBox(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.mode_groupbox.sizePolicy().hasHeightForWidth())
        self.mode_groupbox.setSizePolicy(sizePolicy)
        self.mode_groupbox.setMinimumSize(QtCore.QSize(200, 0))
        self.mode_groupbox.setObjectName("mode_groupbox")
        self.gridLayout_2 = QtWidgets.QGridLayout(self.mode_groupbox)
        self.gridLayout_2.setObjectName("gridLayout_2")
        self.label = QtWidgets.QLabel(self.mode_groupbox)
        self.label.setObjectName("label")
        self.gridLayout_2.addWidget(self.label, 2, 0, 1, 1)
        self.totp_radiobutton = QtWidgets.QRadioButton(self.mode_groupbox)
        self.totp_radiobutton.setObjectName("totp_radiobutton")
        self.gridLayout_2.addWidget(self.totp_radiobutton, 1, 0, 1, 1)
        self.htop_radiobutton = QtWidgets.QRadioButton(self.mode_groupbox)
        self.htop_radiobutton.setChecked(True)
        self.htop_radiobutton.setObjectName("htop_radiobutton")
        self.gridLayout_2.addWidget(self.htop_radiobutton, 0, 0, 1, 1)
        self.totp_algo_combobox = QtWidgets.QComboBox(self.mode_groupbox)
        self.totp_algo_combobox.setEnabled(False)
        self.totp_algo_combobox.setObjectName("totp_algo_combobox")
        self.totp_algo_combobox.addItem("")
        self.totp_algo_combobox.addItem("")
        self.totp_algo_combobox.addItem("")
        self.gridLayout_2.addWidget(self.totp_algo_combobox, 2, 1, 1, 1)
        self.digits_spinbox = QtWidgets.QSpinBox(self.mode_groupbox)
        self.digits_spinbox.setMinimum(6)
        self.digits_spinbox.setMaximum(8)
        self.digits_spinbox.setProperty("value", 6)
        self.digits_spinbox.setObjectName("digits_spinbox")
        self.gridLayout_2.addWidget(self.digits_spinbox, 3, 1, 1, 1)
        self.label_4 = QtWidgets.QLabel(self.mode_groupbox)
        self.label_4.setObjectName("label_4")
        self.gridLayout_2.addWidget(self.label_4, 3, 0, 1, 1)
        self.gridLayout.addWidget(self.mode_groupbox, 1, 1, 1, 1)
        self.widget = QtWidgets.QWidget(self.centralwidget)
        self.widget.setObjectName("widget")
        self.horizontalLayout_3 = QtWidgets.QHBoxLayout(self.widget)
        self.horizontalLayout_3.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout_3.setObjectName("horizontalLayout_3")
        self.apply_pushbutton = QtWidgets.QPushButton(self.widget)
        icon = QtGui.QIcon.fromTheme("document-save")
        self.apply_pushbutton.setIcon(icon)
        self.apply_pushbutton.setObjectName("apply_pushbutton")
        self.horizontalLayout_3.addWidget(self.apply_pushbutton)
        self.reset_pushbutton = QtWidgets.QPushButton(self.widget)
        icon = QtGui.QIcon.fromTheme("gtk-refresh")
        self.reset_pushbutton.setIcon(icon)
        self.reset_pushbutton.setObjectName("reset_pushbutton")
        self.horizontalLayout_3.addWidget(self.reset_pushbutton)
        self.add_pushbutton = QtWidgets.QPushButton(self.widget)
        icon = QtGui.QIcon.fromTheme("list-add")
        self.add_pushbutton.setIcon(icon)
        self.add_pushbutton.setObjectName("add_pushbutton")
        self.horizontalLayout_3.addWidget(self.add_pushbutton)
        self.remove_pushbutton = QtWidgets.QPushButton(self.widget)
        icon = QtGui.QIcon.fromTheme("list-remove")
        self.remove_pushbutton.setIcon(icon)
        self.remove_pushbutton.setObjectName("remove_pushbutton")
        self.horizontalLayout_3.addWidget(self.remove_pushbutton)
        self.gridLayout.addWidget(self.widget, 3, 1, 1, 1)
        self.generate_otp_pushbutton = QtWidgets.QPushButton(self.centralwidget)
        icon = QtGui.QIcon.fromTheme("dialog-password")
        self.generate_otp_pushbutton.setIcon(icon)
        self.generate_otp_pushbutton.setObjectName("generate_otp_pushbutton")
        self.gridLayout.addWidget(self.generate_otp_pushbutton, 3, 0, 1, 1)
        self.horizontalLayout.addLayout(self.gridLayout)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 638, 18))
        self.menubar.setObjectName("menubar")
        self.menuHelp = QtWidgets.QMenu(self.menubar)
        self.menuHelp.setObjectName("menuHelp")
        self.menuFile = QtWidgets.QMenu(self.menubar)
        self.menuFile.setObjectName("menuFile")
        MainWindow.setMenuBar(self.menubar)
        self.actionAbout = QtWidgets.QAction(MainWindow)
        self.actionAbout.setObjectName("actionAbout")
        self.actionExit = QtWidgets.QAction(MainWindow)
        self.actionExit.setObjectName("actionExit")
        self.actionSettings = QtWidgets.QAction(MainWindow)
        self.actionSettings.setObjectName("actionSettings")
        self.menuHelp.addAction(self.actionAbout)
        self.menuFile.addAction(self.actionExit)
        self.menubar.addAction(self.menuFile.menuAction())
        self.menubar.addAction(self.menuHelp.menuAction())
        self.label.setBuddy(self.totp_algo_combobox)
        self.label_4.setBuddy(self.digits_spinbox)

        self.retranslateUi(MainWindow)
        self.htop_radiobutton.toggled['bool'].connect(self.hotp_counter_spinbox.setEnabled)
        self.totp_radiobutton.toggled['bool'].connect(self.totp_algo_combobox.setEnabled)
        self.totp_radiobutton.toggled['bool'].connect(self.totp_start_time_lineedit.setEnabled)
        self.totp_radiobutton.toggled['bool'].connect(self.totp_now_lineedit.setEnabled)
        self.totp_radiobutton.toggled['bool'].connect(self.totp_time_step_spinbox.setEnabled)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "Two Factor Keeper"))
        self.advanced_groupbox.setTitle(_translate("MainWindow", "Advanced options"))
        self.totp_start_time_lineedit.setText(_translate("MainWindow", "1970-01-01 00:00:00 UTC"))
        self.label_5.setText(_translate("MainWindow", "HOTP counter value:"))
        self.label_7.setText(_translate("MainWindow", "TOTP start time (GNU date style):"))
        self.label_8.setText(_translate("MainWindow", "TOTP \'now\' time (GNU date style):"))
        self.label_6.setText(_translate("MainWindow", "TOTP time step size (seconds):"))
        self.totp_now_lineedit.setText(_translate("MainWindow", "now"))
        self.general_groupbox.setTitle(_translate("MainWindow", "General"))
        self.label_2.setText(_translate("MainWindow", "Name:"))
        self.base32_encoded_checkbox.setText(_translate("MainWindow", "Key/secret is base32 encoded"))
        self.label_3.setText(_translate("MainWindow", "Key/secret:"))
        self.mode_groupbox.setTitle(_translate("MainWindow", "Mode and basic options"))
        self.label.setText(_translate("MainWindow", "TOTP algorithm:"))
        self.totp_radiobutton.setText(_translate("MainWindow", "TOTP"))
        self.htop_radiobutton.setText(_translate("MainWindow", "HOTP"))
        self.totp_algo_combobox.setItemText(0, _translate("MainWindow", "sha1"))
        self.totp_algo_combobox.setItemText(1, _translate("MainWindow", "sha256"))
        self.totp_algo_combobox.setItemText(2, _translate("MainWindow", "sha512"))
        self.label_4.setText(_translate("MainWindow", "Digits:"))
        self.apply_pushbutton.setText(_translate("MainWindow", "Apply"))
        self.reset_pushbutton.setText(_translate("MainWindow", "Reset"))
        self.add_pushbutton.setText(_translate("MainWindow", "Add"))
        self.remove_pushbutton.setText(_translate("MainWindow", "Remove"))
        self.generate_otp_pushbutton.setText(_translate("MainWindow", "Generate OTP"))
        self.menuHelp.setTitle(_translate("MainWindow", "He&lp"))
        self.menuFile.setTitle(_translate("MainWindow", "File"))
        self.actionAbout.setText(_translate("MainWindow", "&About"))
        self.actionExit.setText(_translate("MainWindow", "&Quit"))
        self.actionSettings.setText(_translate("MainWindow", "Settings"))

